File: tools/pgegen/elf.py
import functools
import os
import typing


def read_nt_string(fp):
    s = b''
    while not s.endswith(b'\0'):
        s += fp.read(1)
    return s[:-1].decode()


class ProgramHeader:
    def __init__(self, eh):
        self.p_type = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        if eh.intwidth == 2:
            self.p_flags = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.p_offset = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.p_vaddr = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.p_paddr = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.p_filesz = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.p_memsz = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        if eh.intwidth == 1:
            self.p_flags = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.p_align = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)

    def __repr__(self):
        return f'ProgramHeader(0x{self.p_vaddr:08x})'


class SectionHeader:
    def __init__(self, eh):
        self.sh_name = None
        self.sh_name_off = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.sh_type = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.sh_flags = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.sh_addr = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.sh_offset = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.sh_size = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.sh_link = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.sh_info = int.from_bytes(eh.read(4), eh.endianness, signed=False)
        self.sh_addralign = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        self.sh_entsize = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)

    def __repr__(self):
        if self.sh_name is None:
            return f'SectionHeader(anon@0x{self.sh_addr:08x})'
        else:
            return f'SectionHeader({self.sh_name}@0x{self.sh_addr:08x})'


@functools.total_ordering
class Symbol:
    def __init__(self):
        self.st_name = None
        self.st_sh = None
        self.st_name_off = 0
        self.st_value = 0
        self.st_size = 0
        self.st_info = 0
        self.st_other = 0
        self.st_shndx = -1

    def __repr__(self):
        if self.st_name is None:
            return f'Symbol(anon0x{self.st_value})'
        else:
            return f'Symbol({self.st_name}@0x{self.st_value:08x})'

    def __eq__(self, other):
        if isinstance(other, str):
            return self.st_name == other
        if isinstance(other, self.__class__):
            return self.st_name == other.st_name and self.st_value == other.st_value
        if isinstance(other, int):
            return self.st_value == other
        return NotImplemented

    def __lt__(self, other):
        if isinstance(other, self.__class__):
            return self.st_value < other.st_value
        if isinstance(other, int):
            return self.st_value < other
        return NotImplemented

    @classmethod
    def from_eh(cls, eh):
        self = cls()
        self.st_name_off = int.from_bytes(eh.read(4 * eh.intwidth), eh.endianness, signed=False)
        if eh.intwidth == 1:
            self.st_value = int.from_bytes(eh.read(4), eh.endianness, signed=False)
            self.st_size = int.from_bytes(eh.read(4), eh.endianness, signed=False)
            self.st_info = int.from_bytes(eh.read(1), eh.endianness, signed=False)
            self.st_other = int.from_bytes(eh.read(1), eh.endianness, signed=False)
            self.st_shndx = int.from_bytes(eh.read(2), eh.endianness, signed=False)
        else:
            self.st_info = int.from_bytes(eh.read(1), eh.endianness, signed=False)
            self.st_other = int.from_bytes(eh.read(1), eh.endianness, signed=False)
            self.st_shndx = int.from_bytes(eh.read(2), eh.endianness, signed=False)
            self.st_value = int.from_bytes(eh.read(8), eh.endianness, signed=False)
            self.st_size = int.from_bytes(eh.read(8), eh.endianness, signed=False)
        return self

    def __bool__(self):
        return self.st_name is not None


class Elf:
    # Only for ARM32LEv1
    def __init__(self, fp):
        self._fp = fp
        print('Reading ELF file:', fp.name, '... ', end='', flush=True)
        assert self._fp.read(4) == b'\x7fELF'
        self.intwidth = int.from_bytes(self._fp.read(1), 'little', signed=False)
        self.endianness = 'little' if self._fp.read(1) == b'\x01' else 'big'
        self.version = int.from_bytes(self._fp.read(1), 'little', signed=False)
        self.abi = int.from_bytes(self._fp.read(1), 'little', signed=False)
        self.abiver = int.from_bytes(self._fp.read(1), 'little', signed=False)
        self._fp.seek(16)
        self.obj_filetype = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.machine = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.eversion = int.from_bytes(self._fp.read(4), self.endianness, signed=False)
        self.e_entry = int.from_bytes(self._fp.read(self.intwidth * 4), self.endianness, signed=False)
        self.e_phoff = int.from_bytes(self._fp.read(self.intwidth * 4), self.endianness, signed=False)
        self.e_shoff = int.from_bytes(self._fp.read(self.intwidth * 4), self.endianness, signed=False)
        self.e_flags = int.from_bytes(self._fp.read(4), self.endianness, signed=False)
        self.e_ehsize = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.e_phentsize = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.e_phnum = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.e_shentsize = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.e_shnum = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        self.e_shstrndx = int.from_bytes(self._fp.read(2), self.endianness, signed=False)
        print('done')

        print('Reading program headers... ', end='', flush=True)
        self._fp.seek(self.e_phoff)
        self.program_headers: typing.List[ProgramHeader] = [ProgramHeader(self) for i in range(self.e_phnum)]
        print('done')

        print('Reading section headers... ', end='', flush=True)
        self._fp.seek(self.e_shoff)
        self.section_headers: typing.List[SectionHeader] = [SectionHeader(self) for i in range(self.e_shnum)]
        self.shstrtab = self.section_headers[self.e_shstrndx]
        print('done')

        print('Reading symbols and strings... ', end='', flush=True)
        self.strings: typing.Dict[bool, typing.Dict[int, str]] = {}
        self.symbols: typing.List[Symbol] = []
        strconcat = ''
        shstrconcat = ''
        for i, strtab in enumerate(self.section_headers):
            if strtab.sh_type == 3:  # string table
                self._fp.seek(strtab.sh_offset)
                is_shstr = i == self.e_shstrndx
                self.strings[is_shstr] = {}
                while self._fp.tell() < strtab.sh_offset + strtab.sh_size:
                    off = self._fp.tell() - strtab.sh_offset
                    s = read_nt_string(self._fp)
                    if is_shstr:
                        shstrconcat += s + '\0'
                    else:
                        strconcat += s + '\0'
                    self.strings[is_shstr][off] = s
            elif strtab.sh_type == 2:  # symbol table
                self._fp.seek(strtab.sh_offset)
                while self._fp.tell() < strtab.sh_offset + strtab.sh_size:
                    self.symbols.append(Symbol.from_eh(self))
        print('done')

        print('Assigning section names... ', end='', flush=True)
        for sec in self.section_headers:
            s = shstrconcat[sec.sh_name_off:]
            sec.sh_name = s[:s.find('\0')]
        print('done')

        print('Assigning symbol names... ', end='', flush=True)
        for i, sym in enumerate(self.symbols):
            s = strconcat[sym.st_name_off:]
            sym.st_name = s[:s.find('\0')]
            if sym.st_shndx < 0xFF00:
                sym.st_sh = self.section_headers[sym.st_shndx]
        self.symbols_sorted = sorted(self.symbols)
        print('done')

    def __enter__(self):
        return self

    def close(self):
        self._fp.close()
        self._fp = None

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def get_sym(self, name) -> typing.Union[None, Symbol]:
        for sym in self.symbols:
            if sym == name:
                return sym
        # print(f'WARNING: symbol {name} not found, returning dummy object')
        return Symbol()

    def get_offset(self, name):
        sym = self.get_sym(name)
        if sym is not None:
            return sym.st_value
        # print(f'WARNING: symbol {name} not found, returning invalid offset')
        return -1

    @classmethod
    def from_filename(cls, filename):
        return cls(open(filename, 'rb'))

    def read(self, size=-1):
        return self._fp.read(size)

    def seek(self, target, whence=os.SEEK_SET):
        return self._fp.seek(target, whence)

File: tools/pgegen/elf.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "elf.h"
#include "util.h"

static char * strtab = NULL;
static char * shstrtab = NULL;
static Elf32_Sym * symbols = NULL;
static Elf32_Shdr * sectionHeaders = NULL;
static Elf32_Phdr * programHeaders = NULL;
static Elf32_Ehdr elfHeader = {};
static int initialized = 0;
unsigned int nSymbols = 0;
static FILE * elfFile = NULL;

static const char ElfMagic[4] = {'\x7F', 'E', 'L', 'F'};

#ifdef _MSC_VER
#define DEBUG_MSG(fmt, ...) // fprintf(stderr, fmt, __VA_ARGS__)
#else
#define DEBUG_MSG(fmt, ...) // fprintf(stderr, fmt, ##__VA_ARGS__)
#endif

static void ReadElfHeader(void)
{
    DEBUG_MSG("ReadElfHeader\n");

    fseek(elfFile, 0, SEEK_SET);
    fread(&elfHeader, sizeof(elfHeader), 1, elfFile);
    if (memcmp(elfHeader.e_ident, ElfMagic, sizeof(ElfMagic)) != 0) {
        FATAL_ERROR("malformed ELF header\n");
    }
    initialized = 1;
}

static void ReadProgramHeaders(void)
{
    DEBUG_MSG("ReadProgramHeaders\n");

    if (!initialized) {
        FATAL_ERROR("ELF not initialized\n");
    }
    programHeaders = malloc(elfHeader.e_phentsize * elfHeader.e_phnum);
    if (programHeaders == NULL) {
        FATAL_ERROR("failed to allocate programs table\n");
    }
    fseek(elfFile, elfHeader.e_phoff, SEEK_SET);
    fread(programHeaders, elfHeader.e_phentsize, elfHeader.e_phnum, elfFile);
}

static void ReadSectionHeaders(void)
{
    DEBUG_MSG("ReadSectionHeaders\n");

    if (!initialized) {
        FATAL_ERROR("ELF not initialized\n");
    }
    sectionHeaders = malloc(elfHeader.e_shentsize * elfHeader.e_shnum);
    if (sectionHeaders == NULL) {
        FATAL_ERROR("failed to allocate section table\n");
    }
    fseek(elfFile, elfHeader.e_shoff, SEEK_SET);
    fread(sectionHeaders, elfHeader.e_shentsize, elfHeader.e_shnum, elfFile);
}

static char * read_strings(Elf32_Shdr * section)
{
    DEBUG_MSG("read_strings\n"
              "offset: %X, size: %X\n", section->sh_offset, section->sh_size);

    char * dest = malloc(section->sh_size);
    if (dest == NULL) {
        FATAL_ERROR("failed to allocate string table\n");
    }
    fseek(elfFile, section->sh_offset, SEEK_SET);
    fread(dest, 1, section->sh_size, elfFile);
    return dest;
}

static void ReadStringTables(void)
{
    DEBUG_MSG("ReadStringTables\n");

    if (sectionHeaders == NULL) {
        FATAL_ERROR("need to read section headers first\n");
    }
    shstrtab = read_strings(&sectionHeaders[elfHeader.e_shstrndx]);
    for (int i = 0; i < elfHeader.e_shnum; i++)
    {
        if (i == elfHeader.e_shstrndx) continue;
        if (sectionHeaders[i].sh_type == SHT_STRTAB) {
            strtab = read_strings(&sectionHeaders[i]);
            break;
        }
    }
    if (strtab == NULL) {
        FATAL_ERROR("failed to find strings table\n");
    }
}

static int symcmp(const void * a, const void * b)
{
    const Elf32_Sym * aa = (const Elf32_Sym *)a;
    const Elf32_Sym * bb = (const Elf32_Sym *)b;
    return aa->st_value > bb->st_value;
}

static void ReadSymbols(void)
{
    DEBUG_MSG("ReadSymbols\n");

    if (sectionHeaders == NULL) {
        FATAL_ERROR("need to read section headers first\n");
    }
    for (int i = 0; i < elfHeader.e_shnum; i++)
    {
        if (sectionHeaders[i].sh_type == SHT_SYMTAB) {
            nSymbols = sectionHeaders[i].sh_size / sizeof(Elf32_Sym);
            DEBUG_MSG("nSymbols: %d\n", nSymbols);
            symbols = malloc(sectionHeaders[i].sh_size);
            if (symbols == NULL) {
                FATAL_ERROR("failed to allocate symbol table\n");
            }
            fseek(elfFile, sectionHeaders[i].sh_offset, SEEK_SET);
            fread(symbols, sizeof(Elf32_Sym), nSymbols, elfFile);
            msort(symbols, nSymbols, sizeof(Elf32_Sym), symcmp);
            break;
        }
    }
    if (symbols == NULL) {
        FATAL_ERROR("failed to find symbol table\n");
    }
}

Elf32_Sym * GetSymbol(unsigned int st_idx)
{
    DEBUG_MSG("GetSymbol\n");

    if (symbols == NULL) {
        FATAL_ERROR("must read symbols first\n");
    }
    if (st_idx >= nSymbols) {
        FATAL_ERROR("symbol index out of array\n");
    }
    return &symbols[st_idx];
}

Elf32_Sym * GetSymbolOrNullByName(const char * name) {
    DEBUG_MSG("GetSymbolByName\n");

    if (symbols == NULL) {
        FATAL_ERROR("must read symbols first\n");
    }
    for (unsigned int i = 0; i < nSymbols; i++) {
        const char *checkName = strtab + symbols[i].st_name;
        if (strcmp(checkName, name) == 0) {
            return &symbols[i];
        }
    }
    return NULL;
}

Elf32_Sym * GetSymbolByName(const char * name)
{
    Elf32_Sym *symbol = GetSymbolOrNullByName(name);
    if (symbol == NULL)
        FATAL_ERROR("Cannot find symbol %s\n", name);
    return symbol;
}

Elf32_Shdr * GetSectionHeader(unsigned int sh_idx)
{
    DEBUG_MSG("GetSectionHeader\n");

    if ((sh_idx & 0xFF00) == 0xFF00) {  // reserved indices
        return NULL;
    }
    if (sh_idx >= elfHeader.e_shnum) {
        FATAL_ERROR("section index out of array\n");
    }
    return &sectionHeaders[sh_idx];
}

Elf32_Phdr * GetProgramHeader(unsigned int ph_idx)
{
    DEBUG_MSG("GetProgramHeader\n");

    if (ph_idx >= elfHeader.e_phnum) {
        FATAL_ERROR("program index out of array\n");
    }
    return &programHeaders[ph_idx];
}

const char * GetSymbolName(Elf32_Sym * symbol)
{
    DEBUG_MSG("GetSymbolName\n");

    if (strtab == NULL) {
        FATAL_ERROR("must read strings first\n");
    }
    return strtab + symbol->st_name;
}

const char * GetSectionName(Elf32_Shdr * section)
{
    DEBUG_MSG("GetSectionName\n");

    if (shstrtab == NULL) {
        FATAL_ERROR("must read strings first\n");
    }
    return shstrtab + section->sh_name;
}

void InitElf(FILE * file)
{
    elfFile = file;

    ReadElfHeader();
    ReadProgramHeaders();
    ReadSectionHeaders();
    ReadStringTables();
    ReadSymbols();
}

void DestroyResources(void)
{
    DEBUG_MSG("DestroyResources\n");

    free(strtab);
    strtab = NULL;
    free(shstrtab);
    shstrtab = NULL;
    free(programHeaders);
    programHeaders = NULL;
    free(sectionHeaders);
    sectionHeaders = NULL;
    free(symbols);
    symbols = NULL;
    nSymbols = 0;
    elfHeader = (Elf32_Ehdr){};
    initialized = 0;
    elfFile = NULL;
}
